### DISCUSSION TITLE
Upgrading from 0.5 -> 0.7 (namely the `Component` trait)

### DISCUSSION THREAD

**Discussion by zmitchell** [Ogeon/palette#360] [Q&A]
I recently started working on a codebase that used 0.5 and I'm going through and updating the dependencies, but I'm having trouble understanding the upgrade path from 0.5 -> 0.7.

The `Component` trait was implemented for float and integer types, but now the `Stimulus` trait appears to only be implemented for integer types. If I previously had an interface like this:
```rust
/// A short-hand constructor for `Rgb::new`.
pub fn rgb<T>(r: T, g: T, b: T) -> Rgb<T>
where
    T: Component,
{
    srgb(r, g, b)
}
```

Would the new interface look like this?
```rust
/// A short-hand constructor for `Rgb::new`.
pub fn rgb<T, S>(r: T, g: T, b: T) -> Rgb<S>
where
    T: IntoStimulus<S>,
{
    srgb(r, g, b)
}
```

Does this mean that color types can no longer hold floats?

**Comment by Ogeon** [comment]
Hi, they are implemented for float via a blanket implementation for anything real-ish:

````rust
impl<T> Stimulus for T
where
    T: Real + One + Zero,
{
    #[inline]
    fn max_intensity() -> Self {
        Self::one()
    }
}
````

You can see it at the bottom of this page: https://docs.rs/palette/0.7.3/palette/stimulus/trait.Stimulus.html. Your example should work fine _if you need to convert the values from one type to another_. Another change since 0.5 is that `Rgb` and all other color types (if I don't forget any) no longer require any specific trait for being constructed. Only for the operations that need them. You can see here that the trait bounds on the type, itself, are removed: https://docs.rs/palette/0.7.3/palette/rgb/struct.Rgb.html.

It may also be worth looking through the release announcements, if you haven't. They highlight most of the major changes:
* https://ogeon.github.io/2021/07/12/palette-0.6.0.html (some of this, like gradients, is no longer relevant)
* https://ogeon.github.io/2023/04/10/palette-0.7.0.html
* https://ogeon.github.io/2023/05/21/palette-0.7.2.html
* https://ogeon.github.io/2023/08/10/palette-0.7.3.html

I hope this was the answer you were looking for. Don't hesitate to ask about anything else that you encounter!
